File: go/src/leetcode/36.valid-sudoku.go
/*
 * @lc app=leetcode id=36 lang=golang
 *
 * [36] Valid Sudoku
 */

package leetcode

func isValidSudoku(board [][]byte) bool {
	boxs := [3][3][]byte{}
	rows := [9][]byte{}
	columns := [9][]byte{}

	dot := []byte(".")[0]

	for iRow, rowCells := range board {
		for iColumn, cell := range rowCells {
			if cell != dot {
				// same row
				for _, vInRow := range rows[iRow] {
					if vInRow == cell {
						return false
					}
				}
				rows[iRow] = append(rows[iRow], cell)
				// same column
				for _, vInColumn := range columns[iColumn] {
					if vInColumn == cell {
						return false
					}
				}
				columns[iColumn] = append(columns[iColumn], cell)
				// same box
				x := iRow / 3
				y := iColumn / 3
				for _, vInBox := range boxs[x][y] {
					if vInBox == cell {
						return false
					}
				}
				boxs[x][y] = append(boxs[x][y], cell)
			}
		}
	}

	return true
}

// func main() {

// 	s := [][]byte{
// 		[]byte("53..7...."),
// 		[]byte("6..195..."),
// 		[]byte(".98....6."),
// 		[]byte("8...6...3"),
// 		[]byte("4..8.3..1"),
// 		[]byte("7...2...6"),
// 		[]byte(".6....28."),
// 		[]byte("...419..5"),
// 		[]byte("....8..79"),
// 	}
// 	print(isValidSudoku(s))
// }

File: archive/python/36.valid-sudoku.py
#
# @lc app=leetcode id=36 lang=python3
#
# [36] Valid Sudoku
#
import collections


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = collections.defaultdict(set)
        columns = collections.defaultdict(set)
        boxs = collections.defaultdict(set)

        for i_row, row_cells in enumerate(board):
            for i_column, cell in enumerate(row_cells):
                if cell != ".":
                    if cell in rows[i_row] or cell in columns[i_column] or cell in boxs[(i_row//3, i_column//3)]:
                        return False
                    rows[i_row].add(cell)
                    columns[i_column].add(cell)
                    boxs[(i_row//3, i_column//3)].add(cell)
        return True
